### PR TITLE
Add GPT2ForSequenceClassification based on DialogRPT

### LINKED ISSUE #7493: Sharing Microsoft's DialogRPT (new dialog ranking model)
# 🌟 New model addition

## Model description

Thanks for the awesome work!

[DialogRPT](https://github.com/golsun/DialogRPT) (Dialog Ranking Pretrained Transformers) is a set of GPT-2 based dialogue ranking models recently released with an [EMNLP paper](https://arxiv.org/abs/2009.06978) by Microsoft Research. It's a follow-up work of [DialoGPT](https://huggingface.co/transformers/model_doc/dialogpt.html) (thanks for hosting it!)
The architecture is pretty simple: a `GPT2Model` followed by a `torch.nn.Linear(n_embd, 1, bias=False)`, and implemented based on a [previous HuggingFace commit](https://github.com/huggingface/transformers/commit/4d456542e9d381090f9a00b2bcc5a4cb07f6f3f7)
At first, I'm trying to create a model card for it, but then realized that it seems there's no existing model architecture in HuggingFace is compatible with DialogRPT. I noticed a lot of BERT-based sequence classification models, but ours is GPT-2 based. 

If there's a simple fix (or I missed something) please let me know!
If implementation in modeling_gpt2.py is necessary, I'm also glad to help!

## Open source status

* [x] the model implementation is available: (https://github.com/golsun/DialogRPT)
* [x] the model weights are available: (https://github.com/golsun/DialogRPT)
* [x] who are the authors: @golsun @dreasysnail

### PR DESCRIPTION
**Pull request by LysandreJik**
# What does this PR do?

This PR implements `GPT2ForSequenceClassification` in order to support DialogRPT.
Closes https://github.com/huggingface/transformers/issues/7493.

`GPT2ForSequenceClassification` uses the last token in order to do the classification, as other causal models (e.g. GPT-1) do.

Since it does classification on the last token, it requires to know the position of the last token. If a `pad_token_id` is defined in the configuration, it finds the last token that is not a pad token in each row. If no `pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot guess the pad tokens when `inputs_embeds` are passed instead of `input_ids`, it does the same (take the last value in each row of the batch).

Here's how to replicate the results shown on the [original implementation](https://github.com/golsun/DialogRPT#use-rankers-only):

```py
from transformers import GPT2Tokenizer, GPT2ForSequenceClassification
import torch

tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
model = GPT2ForSequenceClassification.from_pretrained("directory_where_pth_config_are_saved")

model_input = tokenizer.encode("I love NLP!<|endoftext|>Here’s a free textbook (URL) in case anyone needs it.", return_tensors="pt")
result = model(model_input, return_dict=True)
final_output = torch.sigmoid(result.logits)

print(final_output)
# tensor([[0.6129]], grad_fn=<SigmoidBackward>)
```

Once this PR is merged I'll open two "Good first issues":

- Implement `GPT2ForSequenceClassification` in TF2
- Implement sequence classification models for other causal transformers

### CODE REVIEW

File: src/transformers/modeling_gpt2.py [hunk 1]
```diff
@@ -946,3 +946,111 @@ def forward(
...
+    GPT2_START_DOCSTRING,
+)
+class GPT2ForSequenceClassification(GPT2PreTrainedModel):
+    authorized_missing_keys = [r"h\.\d+\.attn\.masked_bias", r"lm_head\.weight"]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.transformer = GPT2Model(config)
+        self.score = nn.Linear(config.n_embd, self.num_labels, bias=False)
```
Comment (sgugger) [inline src/transformers/modeling_gpt2.py:963]: Other models use two layers for the classification head. We sure we want only one?

File: src/transformers/modeling_gpt2.py [hunk 2]
```diff
@@ -946,3 +946,111 @@ def forward(
...
+            batch_size, sequence_length = inputs_embeds.shape[:2]
+
+        assert (
+            self.config.pad_token_id is not None or batch_size == 1
+        ), "Cannot handle batch sizes > 1 if no padding token is defined."
+        if self.config.pad_token_id is None:
+            sequence_lengths = -1
+        else:
+            if input_ids is not None:
+                sequence_lengths = torch.ne(input_ids, self.config.pad_token_id).sum(-1) - 1
```
Comment (sgugger) [inline src/transformers/modeling_gpt2.py:1026]: This assumes padding is done on the right. Are we sure it's always going to be the case? Should it be in the model documentation?

File: tests/test_modeling_gpt2.py [hunk 3]
```diff
@@ -332,7 +346,11 @@ def prepare_config_and_inputs_for_common(self):
...
 
-    all_model_classes = (GPT2Model, GPT2LMHeadModel, GPT2DoubleHeadsModel) if is_torch_available() else ()
+    all_model_classes = (
+        (GPT2Model, GPT2LMHeadModel, GPT2DoubleHeadsModel, GPT2ForSequenceClassification)
+        if is_torch_available()
+        else ()
+    )
     all_generative_model_classes = (
         (GPT2LMHeadModel, GPT2DoubleHeadsModel) if is_torch_available() else ()
     )  # TODO (PVP): Add Double HeadsModel when generate() function is changed accordingly
```
Comment (sgugger) [inline tests/test_modeling_gpt2.py:356]: Unrelated, is this comment still relevant?

**Review by sgugger — APPROVED** [review]
Looks good to me with a few questions.
Also, I think this special behavior should be documented. It can all be added in the `@add_start_docstring` before the class.

File: tests/test_modeling_gpt2.py (hunk 3 above)
Comment (patrickvonplaten) [reply tests/test_modeling_gpt2.py:356]: comment can be removed => `GPT2DobuleHeadsModel` works with generate()

File: src/transformers/modeling_gpt2.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_gpt2.py:1026]: Agree - I think we can add a couple of extra lines to make sure `sequence_lengths` is also correct when padding is done from the left.

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/modeling_gpt2.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/modeling_gpt2.py:963]: Adding another layer for the classification head would make this head incompatible with `DialogRPT`, which is the reason this PR is done.

There is no other interesting pre-trained sequence classification GPT-2 model that I know of, but if we hear of one that uses a second layer later on, we can find a way to enable that.

File: src/transformers/modeling_gpt2.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/modeling_gpt2.py:1026]: This can be managed in the tokenizer, but there is currently nothing that can be seen from the model side. I think this should be added cleanly in a second PR, I can take care of it once this PR is merged.
